scripts/delete-account.js

```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env node
+
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/*/
+
+ Usage:
+
+ node scripts/delete-account.js user@example.com
+
+ This script is used to manually delete a user's account,
+ e.g. in response to a GDPR deletion request.  It uses the
+ same config file as key_server.js so should be run from a
+ production instance, and it tries to re-use as much of the
+ actual delete-account route code as possible.
+
+/*/
+
+'use strict'
+
+const readline = require('readline')
+const P = require('../lib/promise')
+const config = require('../config').getProperties()
+const log = require('../lib/log')(config.log.level)
+const Token = require('../lib/tokens')(log, config)
+const error = null
+const mailer = null
+const MockPassword = null
+const mockCustoms = { check: () => {} }
+const push = null
+
+
+var DB = require('../lib/db')(
+  config,
+  log,
+  Token
+)
+
+DB.connect(config[config.db.backend]).then(db => {
+
+  // Bypass password checks.
+  function MockPassword() { }
+  MockPassword.prototype.verifyHash = () { return P.resolve(true) }
+
+  // Bypass TOTP checks.
+  db.totpToken = () => { return P.resolve(false) }
+
+  const signinUtils = require('../lib/routes/utils/signin')(log, config, mockCustoms, db, mailer)
+  const accountDestroyRoute = require('../lib/routes/account')(
+    log,
+    db,
+    mailer,
+    MockPassword,
+    config,
+    mockCustoms,
+    signinUtils,
+    push
+  ).find(r => r.path === '/account/destroy')
+
+
+  P.each(process.argv.slice(1), email => {
+    return db.accountRecord(email).then(account => {
+      console.log("Found account record:")
+      console.log("    uid:", account.uid)
+      console.log("    email:", account.email)
+      const rl = readline.createInterface({
+        input: process.stdin,
+        output: process.stdout,
+      })
+      return new Promise((resolve, reject) => {
+        rl.question('Really delete this account? (y/n) ', answer => {
+          rl.close()
+          if (['y', 'yes'].indexOf(answer.toLowerCase()) === -1) {
+            return reject('Cancelled');
+          }
+          const mockRequest = {
+            payload: {
+              email: email,
+              authPW: 'mock password'
+            }
+          }
+          accountDeleteRoute.handler(mockRequest).then(resolve, reject)
+        })
+      })
+    });
+  }).then(() => {
+    console.log("done!")
+    db.close()
+    // Process won't shut down on its own and I don't know why...
```

Comment:
Does it make a difference if you return the promise from `db.close`?

Reply:
Thanks, tried it just now, but sadly doesn't seem to help

Reply:
Oh wait, it looks weird I know, but what happens if you return the top-level promise too?

```
return DB.connect(
```

I seem to remember having this problem in one of the amplitude scripts and this fixed it. Under the hood I guess node wraps the script in a big IIFE, and so you have to explicitly return the promise even though it looks like your script is running globally or something.